libs/isograph-react/src/core/garbageCollection.ts

```diff
@@ -85,25 +111,47 @@ function recordReachableIdsFromRecord(
         const linkKey = getParentRecordKey(selection, variables ?? {});
         const linkedFieldOrFields = currentRecord[linkKey];
 
-        const ids = [];
+        const links: Link[] = [];
         if (Array.isArray(linkedFieldOrFields)) {
           for (const maybeLink of linkedFieldOrFields) {
             const link = assertLink(maybeLink);
             if (link != null) {
-              ids.push(link.__link);
+              links.push(link);
             }
           }
         } else {
           const link = assertLink(linkedFieldOrFields);
           if (link != null) {
-            ids.push(link.__link);
+            links.push(link);
           }
         }
 
-        for (const nextRecordId of ids) {
-          const nextRecord = store[nextRecordId];
+        let typeStore = selection.concreteType && store[selection.concreteType];
```

Comment:
`??` instead of `&&`. We should never use `&&`, it's unclear, and it allows for bugs. For example, `concreteType` is `""`, then `typeStore === ""`. This also leaks into the type, `typeStore` has type
```
typeStore: "" | {
    [index: string]: StoreRecord | null;
} | null
```